File: src/rust/src/backend/activation.rs
use crate::Activation;
pub struct CPUActivation {
    pub activate: fn(x: &f32) -> f32,
    pub prime: fn(x: &f32) -> f32,
}

impl CPUActivation {
    pub fn from(activation: Activation) -> CPUActivation {
        match activation {
            Activation::Sigmoid => CPUActivation {
                activate: sigmoid,
                prime: sigmoid_prime,
            },
            Activation::Tanh => CPUActivation {
                activate: tanh,
                prime: tanh_prime,
            },
            Activation::Linear => CPUActivation {
                activate: linear,
                prime: linear_prime,
            },
            Activation::Relu => CPUActivation {
                activate: relu,
                prime: relu_prime,
            },
            Activation::Relu6 => CPUActivation {
                activate: relu6,
                prime: relu6_prime,
            },
            Activation::LeakyRelu => CPUActivation {
                activate: leaky_relu,
                prime: leaky_relu_prime,
            },
            Activation::Elu => CPUActivation {
                activate: elu,
                prime: elu_prime,
            },
            Activation::Selu => CPUActivation {
                activate: selu,
                prime: selu_prime,
            },
        }
    }

    pub fn from_option(activation: Option<Activation>) -> Option<CPUActivation> {
        if let Some(activation) = activation {
            Some(CPUActivation::from(activation))
        } else {
            None
        }
    }
}

fn sigmoid(x: &f32) -> f32 {
    return 1.0 / (1.0 + (-x).exp());
}

fn sigmoid_prime(x: &f32) -> f32 {
    return x * (1.0 - x);
}

fn tanh(x: &f32) -> f32 {
    return x.tanh();
}

fn tanh_prime(x: &f32) -> f32 {
    return 1.0 - tanh(x).powi(2);
}

fn linear(x: &f32) -> f32 {
    return *x;
}

fn linear_prime(_x: &f32) -> f32 {
    return 1.0;
}

fn relu(x: &f32) -> f32 {
    return x.max(0.0);
}

fn relu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { 0.0 };
}

fn relu6(x: &f32) -> f32 {
    return x.max(0.0).min(6.0);
}

fn relu6_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { 0.0 };
}

fn leaky_relu(x: &f32) -> f32 {
    return if *x > 0.0 { *x } else { x.max(0.01 * x) };
}

fn leaky_relu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { 0.01 };
}

fn elu(x: &f32) -> f32 {
    return if *x >= 0.0 { *x } else { x.exp() - 1.0 };
}

fn elu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { x.exp() };
}

fn selu(x: &f32) -> f32 {
    return if *x >= 0.0 { *x } else { 1.0507 * (x.exp() - 1.0) };
}

fn selu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { 1.0507 * x.exp() };
}

File: src/backend_gpu/activation.ts
import { Activation } from "../../mod.ts";
import { ActivationError } from "../core/api/error.ts";

export interface GPUActivationFn {
  name: string;
  activate(...values: string[]): string;
  prime(...values: string[]): string;
}

export function setActivation(activation: Activation) {
  switch (activation) {
    case "sigmoid":
      return new Sigmoid();
    case "leakyrelu":
      return new LeakyRelu();
    case "tanh":
      return new Tanh();
    case "relu":
      return new Relu();
    case "relu6":
      return new Relu6();
    case "elu":
      return new Elu();
    case "selu":
      return new Selu();
    case "linear":
      return new Linear();
    default:
      throw new ActivationError(activation);
  }
}

export class Linear implements GPUActivationFn {
  name = "linear";
  activate() {
    return `return weighted_sum`;
  }

  prime() {
    return `return f32(1)`;
  }
}

export class Sigmoid implements GPUActivationFn {
  name = "sigmoid";
  activate() {
    return `return f32(1) / (f32(1) + exp(-weighted_sum))`;
  }

  prime() {
    return `return output * (f32(1) - output)`;
  }
}

export class Tanh implements GPUActivationFn {
  name = "tanh";
  activate() {
    return `return tanh(weighted_sum)`;
  }

  prime() {
    return `return f32(1) - output * output`;
  }
}

export class Relu implements GPUActivationFn {
  name = "relu";
  activate() {
    return `return max(f32(0), weighted_sum)`;
  }

  prime() {
    return `if (weighted_sum <= f32(0)) {
            return f32(0);
        }
        return errror;`;
  }
}

export class Relu6 implements GPUActivationFn {
  name = "relu6";
  activate() {
    return `return min(max(f32(0), weighted_sum), f32(6))`;
  }

  prime() {
    return `if (weighted_sum <= f32(0)) {
            return f32(0);
        }
        if (weighted_sum >= f32(6)) {
            return f32(6);
        }
        return error;`;
  }
}

export class LeakyRelu implements GPUActivationFn {
  name = "leakyrelu";
  activate() {
    return `if (weighted_sum > f32(0)) {
            return weighted_sum;
        }
        return f32(f32(weighted_sum) * 0.01);`;
  }

  prime() {
    return `if (weighted_sum > f32(0)) {
            return error;
        }
        return f32(f32(error) * 0.01);`;
  }
}

export class Elu implements GPUActivationFn {
  name = "elu";
  activate() {
    return `if (weighted_sum > f32(0)) {
            return weighted_sum;
        }
        return f32(exp(weighted_sum) - f32(1));`;
  }

  prime() {
    return `if (weighted_sum > f32(0)) {
            return error;
        }
        return f32(exp(weighted_sum) - f32(1));`;
  }
}

export class Selu implements GPUActivationFn {
  name = "selu";
  activate() {
    return `return f32(weighted_sum) + f32(weighted_sum) * (f32(1) - f32(weighted_sum)) * f32(1.67326)`;
  }

  prime() {
    return `if (weighted_sum > f32(0)) {
            return error;
        }
        return f32(error) * (f32(1) - f32(weighted_sum)) * f32(1.67326);`;
  }
}
